### PR TITLE
Make the Nested Rows plugin reject data other than Array of Objects.

### PR DESCRIPTION
**Pull request by jansiegel**
### Context
This PR adds a check for the Nested Rows plugin, ensuring `data` is provided.

### How has this been tested?
Added a test case.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #8171 

### Affected project(s):
- [x] `handsontable`
- [ ] `@handsontable/angular`
- [ ] `@handsontable/react`
- [ ] `@handsontable/vue`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: src/plugins/nestedRows/data/dataManager.js [hunk 1]
```diff
@@ -96,6 +96,10 @@ class DataManager {
    * @private
    */
   rewriteCache() {
+    if (!this.data?.length) {
```
Comment (wszymanski) [inline src/plugins/nestedRows/data/dataManager.js:99]:
```suggestion
    if (Array.isArray(data) === false || data.length === 0) {
```

In my opinion, it would me more clear, but it's up to you. But, probably it won't be needed after another change.

**Review by wszymanski — CHANGES_REQUESTED** [review]
For me, the below configuration shouldn't result in displaying no rows. I think that there should be `5` start rows and cols.

```
handsontable({
  nestedRows: true
});
```

Please consider to add test for it.

**Comment by jansiegel** [comment]
> For me, the below configuration shouldn't result in displaying no rows. I think that there should be `5` start rows and cols.
> 
> ```
> handsontable({
>   nestedRows: true
> });
> ```
> 
> Please consider to add test for it.

@wszymanski 

After some testing (and realizing that even if the table displayed the default `5x5` it would still crash, as it's generated as an AoA) and consulting the problem with @wojciechczerniak, I changed the approach to the problem:

After the latest change, the plugin will throw an appropriate error and disable itself if the provided data is not Array of Objects. [It's consistent with information in our current documentation.](https://handsontable.com/docs/row-parent-child/#preparing-the-data-source)

File: src/plugins/nestedRows/__tests__/nestedRows.spec.js [hunk 2]
```diff
@@ -29,6 +29,53 @@ describe('NestedRows', () => {
...
+      const errorSpy = jasmine.createSpy('init errors');
+
+      // eslint-disable-next-line no-console
+      const prevError = console.error.bind(console);
+
+      // eslint-disable-next-line no-console
+      console.error = (...args) => {
+        errorSpy(...args);
+        prevError(...args);
+      };
```
Comment (wszymanski) [inline src/plugins/nestedRows/__tests__/nestedRows.spec.js:43]:
```suggestion
      const errorSpy = spyOn(console, 'error');
```

🙂

File: src/plugins/nestedRows/__tests__/nestedRows.spec.js [hunk 3]
```diff
@@ -29,6 +29,53 @@ describe('NestedRows', () => {
...
+
+      handsontable({
+        nestedRows: true
+      });
+
+      expect(errorSpy).toHaveBeenCalledWith(Handsontable.plugins.NestedRows.WRONG_DATA_TYPE_ERROR);
+
+      // eslint-disable-next-line no-console
+      console.error = prevError;
```
Comment (wszymanski) [inline src/plugins/nestedRows/__tests__/nestedRows.spec.js:52]:
```suggestion
```

Related to the above comment with using `spyOn`.

File: src/plugins/nestedRows/__tests__/nestedRows.spec.js [hunk 4]
```diff
@@ -29,6 +29,53 @@ describe('NestedRows', () => {
...
+      const errorSpy = jasmine.createSpy('init errors');
+
+      // eslint-disable-next-line no-console
+      const prevError = console.error.bind(console);
+
+      // eslint-disable-next-line no-console
+      console.error = (...args) => {
+        errorSpy(...args);
+        prevError(...args);
+      };
```
Comment (wszymanski) [inline src/plugins/nestedRows/__tests__/nestedRows.spec.js:65]:
```suggestion
      const errorSpy = spyOn(console, 'error');
```

File: src/plugins/nestedRows/__tests__/nestedRows.spec.js [hunk 5]
```diff
@@ -29,6 +29,53 @@ describe('NestedRows', () => {
...
+
+      handsontable({
+        data: Handsontable.helper.createSpreadsheetData(3, 3),
+        nestedRows: true
+      });
+
+      expect(errorSpy).toHaveBeenCalledWith(Handsontable.plugins.NestedRows.WRONG_DATA_TYPE_ERROR);
+
+      // eslint-disable-next-line no-console
+      console.error = prevError;
```
Comment (wszymanski) [inline src/plugins/nestedRows/__tests__/nestedRows.spec.js:75]:
```suggestion
```

File: src/plugins/nestedRows/nestedRows.js [hunk 6]
```diff
@@ -28,6 +31,14 @@ export class NestedRows extends BasePlugin {
     return PLUGIN_PRIORITY;
   }
 
+  /**
+   * Error message for the wrong data type error.
+   *
+   * @returns {string}
+   */
+  static WRONG_DATA_TYPE_ERROR = 'The Nested Rows plugin requires an Array of Objects as a dataset to be' +
+    ' provided. The plugin has been disabled.';
```
Comment (wszymanski) [inline src/plugins/nestedRows/nestedRows.js:40]: I would define it as constant at top of the file, but it's up to you.

**Review by wszymanski — CHANGES_REQUESTED** [review]

File: src/plugins/nestedRows/nestedRows.js (hunk 6 above)
Comment (jansiegel) [reply src/plugins/nestedRows/nestedRows.js:40]: I was going to, but I figured it would be useful for testing purposes to use a static instead of copy/pasting the error message, but you're probably right 👍

File: src/plugins/nestedRows/__tests__/nestedRows.spec.js (hunk 2 above)
Comment (jansiegel) [reply src/plugins/nestedRows/__tests__/nestedRows.spec.js:43]: 🤦

**Review by wszymanski — DISMISSED** [review]
LGFM 🎉

**Comment by jansiegel** [comment]
@wszymanski As [one of the Core's unit tests was failing](https://github.com/handsontable/handsontable/blob/master/test/unit/Core.unit.js#L56) (it uses an AoA dataset, which makes the `nestedRows` plugin disable itself, thus resetting the index mapper cache), I removed `nestedRows` from it and duplicated the entire test, but with AoO dataset, this time including the `nestedRows` plugin.

tl;dr; I'm afraid I'm going to need another review 😆

**Review by wszymanski — APPROVED** [review]
